File: scripts/harness_validation/repository.py
"""校验 Harness 仓库结构、日期记忆、Skills 元数据与本地链接。"""

from __future__ import annotations

import re
from pathlib import Path
from urllib.parse import unquote

from .context import *  # noqa: F403
from .repository_memory import (
    validate_current_changelog_contract,
    validate_daily_project_memory,
    validate_superseded_release_lifecycle_fragments,
)


def validate_required_files(errors: list[str]) -> None:
    """确认所有规范文档、脚本和门禁入口真实存在。"""
    for relative in REQUIRED_FILES:
        if not (ROOT / relative).is_file():
            fail(errors, f"missing required file: {relative}")


def validate_work_plan_contract(
    errors: list[str],
    plan_path: Path = WORK_PLAN,
    *,
    required: bool = False,
) -> None:
    """校验用户按需创建的精简 Work Plan，并拒绝候选事实与验收结论。"""
    if not plan_path.is_file():
        if required:
            fail(errors, f"missing active Work Plan: {display_path(plan_path)}")
        return

    text = read_text_cached(plan_path)
    if "## Todo" not in text:
        fail(errors, f"active Work Plan has no Todo section: {display_path(plan_path)}")

    heading_pattern = re.compile(
        r"^###\s+(TODO-[A-Z0-9-]+)(.*?)$",
        flags=re.MULTILINE,
    )
    heading_matches = list(heading_pattern.finditer(text))
    todo_ids = [match.group(1) for match in heading_matches]
    if not todo_ids:
        fail(errors, f"active Work Plan has no stable Todo IDs: {display_path(plan_path)}")
    duplicates = sorted(
        todo_id for todo_id in set(todo_ids) if todo_ids.count(todo_id) > 1
    )
    if duplicates:
        fail(
            errors,
            "active Work Plan contains duplicate Todo IDs: " + ", ".join(duplicates),
        )

    for index, match in enumerate(heading_matches):
        todo_id = match.group(1)
        heading_suffix = match.group(2)
        states = re.findall(
            r"[（(](pending|in_progress|blocked|done)[）)]",
            heading_suffix,
        )
        if len(states) != 1:
            fail(
                errors,
                f"Todo {todo_id} heading must carry exactly one explicit state",
            )
        block_end = (
            heading_matches[index + 1].start()
            if index + 1 < len(heading_matches)
            else len(text)
        )
        block = text[match.end() : block_end]
        field_patterns = {
            "expected behavior": r"(?:预期行为|Expected behavior)\s*[：:]",
            "ownership/boundary": r"(?:影响边界|Ownership/Boundary)\s*[：:]",
            "verification": r"(?:完成验证|验证|Verification)\s*[：:]",
        }
        for label, pattern in field_patterns.items():
            if not re.search(pattern, block, flags=re.IGNORECASE):
                fail(errors, f"Todo {todo_id} is missing per-item {label}")

    acceptance_matches = list(
        re.finditer(
            r"^##\s+(?:验证里程碑|完整验收)\b.*$",
            text,
            flags=re.MULTILINE,
        )
    )
    for acceptance in acceptance_matches:
        acceptance_block = text[acceptance.end() :]
        for fragment in ("候选", "`done`", "$desktop-implement-change"):
            if fragment not in acceptance_block:
                fail(
                    errors,
                    f"Work Plan acceptance contract missing in {display_path(plan_path)}: {fragment}",
                )
        if not re.search(r"完整(?:真实| Harness|源树|产物)", acceptance_block):
            fail(
                errors,
                f"Work Plan acceptance lacks a complete real candidate: {display_path(plan_path)}",
            )
        if not re.search(r"模拟|桩|占位|脚手架|开发预览|单段文案", acceptance_block):
            fail(
                errors,
                f"Work Plan acceptance lacks substitute rejection: {display_path(plan_path)}",
            )

    candidate_evidence_patterns = (
        re.compile(
            r"^\s*(?:[-*]\s*)?"
            r"(?:(?:当前)?(?:里程碑|验收|技术验收)?(?:状态|结论)"
            r"|(?:Milestone|Acceptance)\s+(?:status|verdict))"
            r"\s*[：:]\s*`?"
            r"(?:Technically\s+accepted|Milestone\s+accepted|accepted|passed|已验收|通过)\b",
            flags=re.IGNORECASE | re.MULTILINE,
        ),
        re.compile(
            r"^\s*(?:[-*]\s*)?"
            r"(?:(?:发布|候选)(?:状态|结论)|发布就绪|Release\s+readiness)"
            r"\s*[：:]\s*`?(?:ready|已就绪|可发布)\b",
            flags=re.IGNORECASE | re.MULTILINE,
        ),
        re.compile(
            r"^\s*(?:[-*]\s*)?"
            r"(?:milestoneAcceptance|sourceCommit|buildRun|buildMode|performanceStatus|"
            r"runtimeVerification|signingStatus|notarizationStatus|sha256)\s*[：:]",
            flags=re.IGNORECASE | re.MULTILINE,
        ),
        re.compile(
            r"^\s*(?:[-*]\s*)?"
            r"(?:(?:候选|构建|E2E|性能)(?:状态|结论)|Candidate\s+(?:status|verdict))"
            r"\s*[：:]\s*`?(?:pending|rejected|accepted|passed|failed|waived|Unverified)\b",
            flags=re.IGNORECASE | re.MULTILINE,
        ),
    )
    if any(pattern.search(text) for pattern in candidate_evidence_patterns):
        fail(
            errors,
            "active Work Plan must not record candidate evidence or an acceptance/readiness verdict",
        )


def parse_frontmatter(path: Path, errors: list[str]) -> dict[str, str]:
    """解析 Skill 的最小 YAML frontmatter，并拒绝缺失或额外字段。"""
    text = path.read_text(encoding="utf-8")
    match = re.match(r"\A---\n(.*?)\n---\n", text, flags=re.DOTALL)
    if not match:
        fail(errors, f"missing YAML frontmatter: {path.relative_to(ROOT)}")
        return {}

    fields: dict[str, str] = {}
    for line in match.group(1).splitlines():
        key, separator, value = line.partition(":")
        if not separator:
            fail(errors, f"invalid frontmatter line in {path.relative_to(ROOT)}: {line}")
            continue
        fields[key.strip()] = value.strip()
    if set(fields) != {"name", "description"}:
        fail(
            errors,
            f"frontmatter must contain only name/description: {path.relative_to(ROOT)}",
        )
    return fields

def yaml_string(text: str, key: str) -> str | None:
    """从受控 UI 元数据中提取一个双引号字符串字段。"""
    match = re.search(rf'^\s*{re.escape(key)}:\s*"([^"]*)"\s*$', text, re.MULTILINE)
    return match.group(1) if match else None

def validate_skills(errors: list[str]) -> None:
    """校验 Skill 集合、frontmatter、UI 元数据和入口声明保持一致。"""
    actual = {path.name for path in SKILLS_ROOT.iterdir() if path.is_dir()}
    if actual != EXPECTED_SKILLS:
        fail(
            errors,
            f"skill set mismatch: missing={sorted(EXPECTED_SKILLS - actual)}, "
            f"extra={sorted(actual - EXPECTED_SKILLS)}",
        )

    readme = (ROOT / "README.md").read_text(encoding="utf-8")
    agents = (ROOT / "AGENTS.md").read_text(encoding="utf-8")
    for skill in sorted(actual):
        skill_dir = SKILLS_ROOT / skill
        skill_file = skill_dir / "SKILL.md"
        metadata_file = skill_dir / "agents" / "openai.yaml"
        if not skill_file.is_file():
            fail(errors, f"missing SKILL.md: {skill}")
            continue
        if not metadata_file.is_file():
            fail(errors, f"missing agents/openai.yaml: {skill}")
            continue

        fields = parse_frontmatter(skill_file, errors)
        if fields.get("name") != skill:
            fail(errors, f"skill name does not match directory: {skill}")
        if not fields.get("description"):
            fail(errors, f"empty skill description: {skill}")
        if "TODO" in skill_file.read_text(encoding="utf-8"):
            fail(errors, f"unresolved TODO in skill: {skill}")

        metadata = metadata_file.read_text(encoding="utf-8")
        display_name = yaml_string(metadata, "display_name")
        short_description = yaml_string(metadata, "short_description")
        default_prompt = yaml_string(metadata, "default_prompt")
        if not display_name:
            fail(errors, f"missing display_name: {skill}")
        if not short_description or not 25 <= len(short_description) <= 64:
            fail(errors, f"short_description length must be 25-64: {skill}")
        if not default_prompt or f"${skill}" not in default_prompt:
            fail(errors, f"default_prompt must mention ${skill}: {skill}")

        if f"`${skill}`" not in readme and f"`${'$'}{skill}`" not in readme:
            fail(errors, f"README does not declare skill: {skill}")
        if f"${skill}" not in agents:
            fail(errors, f"AGENTS routing does not mention skill: {skill}")

def validate_markdown_links(errors: list[str]) -> None:
    """解析仓库内 Markdown 链接，并拒绝指向不存在本地目标的引用。"""
    link_pattern = re.compile(r"\[[^\]]+\]\(([^)]+)\)")
    markdown_files = sorted(ROOT.rglob("*.md"))
    for path in markdown_files:
        text = path.read_text(encoding="utf-8")
        for raw_target in link_pattern.findall(text):
            target = raw_target.strip().strip("<>")
            if target.startswith(("http://", "https://", "mailto:", "#")):
                continue
            target = unquote(target.split("#", 1)[0])
            if not target:
                continue
            resolved = (path.parent / target).resolve()
            if not resolved.exists():
                fail(
                    errors,
                    f"broken local link in {path.relative_to(ROOT)}: {raw_target}",
                )

File: scripts/harness_validation/repository.mjs
import fs from "node:fs";
import path from "node:path";

import {
  ROOT,
  SKILLS_ROOT,
  fail,
  parseFrontmatter,
  readText,
  relativePath,
  resolveInsideRoot,
  trackedFiles,
} from "./core.mjs";
import { inspectProject as inspectNoPython } from "../../.agents/skills/desktop-implement-change/scripts/check_no_python.mjs";
import { validateDailyProjectMemory } from "./repository_memory.mjs";
import { REQUIRED_ROOT_FILES } from "./repository_required_files.mjs";

/** Harness 当前必须保留的 42 个项目 Skill。 */
export const EXPECTED_SKILLS = new Set([
  "desktop-add-cli-adapter",
  "desktop-add-gui-adapter",
  "desktop-add-gui-autostart",
  "desktop-add-gui-deep-link",
  "desktop-add-gui-dialog",
  "desktop-add-gui-global-shortcut",
  "desktop-add-gui-single-instance",
  "desktop-add-gui-system-locale",
  "desktop-add-gui-system-notifications",
  "desktop-add-gui-system-tray",
  "desktop-add-gui-updater",
  "desktop-add-gui-window-state",
  "desktop-add-mcp-adapter",
  "desktop-add-tui-adapter",
  "desktop-build-rust-release",
  "desktop-build-tauri-local-install",
  "desktop-build-tauri-release",
  "desktop-check-development-environment",
  "desktop-collect-release-artifacts",
  "desktop-configure-git-commits",
  "desktop-curate-harness-memory",
  "desktop-define-product",
  "desktop-extract-i18n-strings",
  "desktop-implement-change",
  "desktop-initialize-rust-project",
  "desktop-instantiate-project",
  "desktop-manage-git-lifecycle",
  "desktop-manage-version",
  "desktop-plan-change",
  "desktop-prepare-cross-platform-release",
  "desktop-prepare-gui-app-identity",
  "desktop-prepare-gui-support-surfaces",
  "desktop-prepare-release",
  "desktop-refactor-code",
  "desktop-rename-project-identity",
  "desktop-run-parallel-worktrees",
  "desktop-summarize-development-history",
  "desktop-test-final-artifact-e2e",
  "desktop-test-gui-initialization-e2e",
  "desktop-upgrade-harness",
  "desktop-verify-delivery",
  "mantine-list-view",
]);

/** 验证固定入口、许可证、事实源和 Node 校验入口存在。 */
export function validateRequiredFiles(errors) {
  for (const relative of REQUIRED_ROOT_FILES) {
    const filePath = path.join(ROOT, relative);
    if (!fs.existsSync(filePath)) {
      fail(errors, `缺少必需文件: ${relative}`);
      continue;
    }
    const stat = fs.lstatSync(filePath);
    if (!stat.isFile() || stat.isSymbolicLink()) {
      fail(errors, `必需路径必须是普通非符号链接文件: ${relative}`);
    }
  }
}

/** 验证 Skill 集合、frontmatter 与元数据入口保持一一对应。 */
export function validateSkills(errors) {
  let entries = [];
  try {
    entries = fs.readdirSync(SKILLS_ROOT, { withFileTypes: true });
  } catch (error) {
    fail(errors, `无法枚举项目 Skills: ${error.message}`);
    return;
  }
  const actual = new Set(entries.filter((entry) => entry.isDirectory()).map((entry) => entry.name));
  let readme = "";
  let agents = "";
  try {
    readme = readText(path.join(ROOT, "README.md"));
    agents = readText(path.join(ROOT, "AGENTS.md"));
  } catch (error) {
    fail(errors, error.message);
  }
  for (const name of EXPECTED_SKILLS) {
    if (!actual.has(name)) fail(errors, `缺少预期 Skill: ${name}`);
  }
  for (const name of actual) {
    if (!EXPECTED_SKILLS.has(name)) fail(errors, `Skills 地图未登记目录: ${name}`);
  }
  for (const name of EXPECTED_SKILLS) {
    const skillFile = path.join(SKILLS_ROOT, name, "SKILL.md");
    const metadataFile = path.join(SKILLS_ROOT, name, "agents", "openai.yaml");
    if (!fs.existsSync(skillFile)) {
      fail(errors, `Skill 缺少 SKILL.md: ${name}`);
      continue;
    }
    try {
      const frontmatter = parseFrontmatter(readText(skillFile));
      if (!frontmatter) {
        fail(errors, `Skill 缺少合法 frontmatter: ${name}`);
      } else {
        const keys = Object.keys(frontmatter).sort();
        if (keys.join("\0") !== ["description", "name"].join("\0")) {
          fail(errors, `Skill frontmatter 只能包含 name/description: ${name}`);
        }
        if (frontmatter.name !== name) fail(errors, `Skill name 与目录不一致: ${name}`);
        if (!frontmatter.description?.trim()) fail(errors, `Skill description 不能为空: ${name}`);
      }
      const skillText = readText(skillFile);
      if (skillText.includes("TODO")) fail(errors, `Skill 含未解决 TODO: ${name}`);
      if (name === "mantine-list-view" && skillText.replaceAll("\r\n", "\n").split("\n").length > 201) {
        fail(errors, `mantine-list-view SKILL.md 超过 200 行`);
      }
    } catch (error) {
      fail(errors, error.message);
    }
    if (!fs.existsSync(metadataFile)) {
      fail(errors, `Skill 缺少 agents/openai.yaml: ${name}`);
    } else {
      try {
        const metadata = readText(metadataFile);
        const yamlString = (key) => new RegExp(`^\\s*${key}:\\s*"([^"]*)"\\s*$`, "mu").exec(metadata)?.[1] ?? null;
        const displayName = yamlString("display_name");
        const shortDescription = yamlString("short_description");
        const defaultPrompt = yamlString("default_prompt");
        if (!displayName) fail(errors, `Skill metadata 缺少 display_name: ${name}`);
        if (!shortDescription || shortDescription.length < 25 || shortDescription.length > 64) {
          fail(errors, `Skill metadata short_description 长度必须为 25-64: ${name}`);
        }
        if (!defaultPrompt?.includes(`$${name}`)) fail(errors, `Skill metadata default_prompt 必须提及 $${name}`);
      } catch (error) {
        fail(errors, error.message);
      }
    }
    if (!readme.includes(`\`${name}\``) && !readme.includes(`\`$${name}\``)) fail(errors, `README 未声明 Skill: ${name}`);
    if (!agents.includes(`$${name}`)) fail(errors, `AGENTS 路由未提及 Skill: ${name}`);
  }
}

/** 校验可选精简 Work Plan 的 Todo 状态、逐项字段和候选事实隔离。 */
export function validateWorkPlanContract(errors, planPath = null, { required = false } = {}) {
  const workPlanDirectory = path.join(ROOT, "docs", "work_plan");
  const selected = planPath ?? fs.readdirSync(workPlanDirectory)
    .filter((name) => /^\d{8}_work_plan\.md$/u.test(name))
    .sort()
    .map((name) => path.join(workPlanDirectory, name))
    .at(-1);
  if (!selected || !fs.existsSync(selected)) {
    if (required) fail(errors, `missing active Work Plan: ${selected ? relativePath(selected) : "docs/work_plan/__missing_latest__.md"}`);
    return;
  }
  const text = readText(selected);
  if (!text.includes("## Todo")) fail(errors, `active Work Plan has no Todo section: ${relativePath(selected)}`);
  const headings = [...text.matchAll(/^###\s+(TODO-[A-Z0-9-]+)(.*?)$/gmu)];
  const todoIds = headings.map((match) => match[1]);
  if (todoIds.length === 0) fail(errors, `active Work Plan has no stable Todo IDs: ${relativePath(selected)}`);
  const duplicates = [...new Set(todoIds.filter((id, index) => todoIds.indexOf(id) !== index))].sort();
  if (duplicates.length > 0) fail(errors, `active Work Plan contains duplicate Todo IDs: ${duplicates.join(", ")}`);
  for (const [index, match] of headings.entries()) {
    const states = [...match[2].matchAll(/[（(](pending|in_progress|blocked|done)[）)]/gu)];
    if (states.length !== 1) fail(errors, `Todo ${match[1]} heading must carry exactly one explicit state`);
    const blockEnd = headings[index + 1]?.index ?? text.length;
    const block = text.slice((match.index ?? 0) + match[0].length, blockEnd);
    const fields = [
      ["expected behavior", /(?:预期行为|Expected behavior)\s*[：:]/iu],
      ["ownership/boundary", /(?:影响边界|Ownership\/Boundary)\s*[：:]/iu],
      ["verification", /(?:完成验证|验证|Verification)\s*[：:]/iu],
    ];
    for (const [label, pattern] of fields) {
      if (!pattern.test(block)) fail(errors, `Todo ${match[1]} is missing per-item ${label}`);
    }
  }
  for (const acceptance of text.matchAll(/^##\s+(?:验证里程碑|完整验收)\b.*$/gmu)) {
    const block = text.slice((acceptance.index ?? 0) + acceptance[0].length);
    for (const fragment of ["候选", "`done`", "$desktop-implement-change"]) {
      if (!block.includes(fragment)) fail(errors, `Work Plan acceptance contract missing in ${relativePath(selected)}: ${fragment}`);
    }
    if (!/完整(?:真实| Harness|源树|产物)/u.test(block)) fail(errors, `Work Plan acceptance lacks a complete real candidate: ${relativePath(selected)}`);
    if (!/模拟|桩|占位|脚手架|开发预览|单段文案/u.test(block)) fail(errors, `Work Plan acceptance lacks substitute rejection: ${relativePath(selected)}`);
  }
  const evidencePatterns = [
    /^\s*(?:[-*]\s*)?(?:(?:当前)?(?:里程碑|验收|技术验收)?(?:状态|结论)|(?:Milestone|Acceptance)\s+(?:status|verdict))\s*[：:]\s*`?(?:Technically\s+accepted|Milestone\s+accepted|accepted|passed|已验收|通过)\b/imu,
    /^\s*(?:[-*]\s*)?(?:(?:发布|候选)(?:状态|结论)|发布就绪|Release\s+readiness)\s*[：:]\s*`?(?:ready|已就绪|可发布)\b/imu,
    /^\s*(?:[-*]\s*)?(?:milestoneAcceptance|sourceCommit|buildRun|buildMode|runtimeVerification|signingStatus|notarizationStatus|sha256)\s*[：:]/imu,
    /^\s*(?:[-*]\s*)?(?:(?:候选|构建|E2E)(?:状态|结论)|Candidate\s+(?:status|verdict))\s*[：:]\s*`?(?:pending|rejected|accepted|passed|failed|waived|Unverified)\b/imu,
  ];
  if (evidencePatterns.some((pattern) => pattern.test(text))) {
    fail(errors, "active Work Plan must not record candidate evidence or an acceptance/readiness verdict");
  }
}

/** 解析本地 Markdown 链接并验证目标存在、未越过仓库根。 */
export function validateMarkdownLinks(errors, files = trackedFiles()) {
  const markdownFiles = files.filter((relative) => relative.endsWith(".md"));
  const linkPattern = /!?\[[^\]]*\]\(([^)]+)\)/gu;
  for (const relative of markdownFiles) {
    const filePath = path.join(ROOT, relative);
    let text;
    try {
      text = readText(filePath);
    } catch (error) {
      fail(errors, error.message);
      continue;
    }
    for (const match of text.matchAll(linkPattern)) {
      let target = match[1].trim();
      if (!target || target.startsWith("#") || /^(?:https?:|mailto:|data:|app:|plugin:|codex:)/iu.test(target)) continue;
      if (target.startsWith("<") && target.endsWith(">")) target = target.slice(1, -1);
      target = target.split("#", 1)[0].split("?", 1)[0];
      if (!target) continue;
      try {
        target = decodeURIComponent(target);
        const resolved = resolveInsideRoot(path.dirname(filePath), target);
        if (!fs.existsSync(resolved)) {
          fail(errors, `失效本地 Markdown 链接: ${relative} -> ${target}`);
        }
      } catch (error) {
        fail(errors, `${relative} 的本地链接无效: ${error.message}`);
      }
    }
  }
}

/** 验证日期记忆索引只指向实际存在的最新快照。 */
export function validateMemoryIndexes(errors) {
  const configurations = [
    ["docs/product_spec", /^(\d{8})_product_spec\.md$/u],
    ["docs/project_status", /^(\d{8})_product_status\.md$/u],
    ["docs/work_plan", /^(\d{8})_work_plan\.md$/u],
    ["docs/adr", /^(\d{8})_ADR\.md$/u],
    ["docs/changelog", /^(\d{8})_CHANGELOG\.md$/u],
  ];
  for (const [directory, pattern] of configurations) {
    const absolute = path.join(ROOT, directory);
    const files = fs.readdirSync(absolute).filter((name) => pattern.test(name)).sort();
    if (files.length !== 1) {
      fail(errors, `${directory} 必须只保留一个日期最新正文，实际 ${files.length} 个`);
      continue;
    }
    const index = readText(path.join(absolute, "README.md"));
    if (!index.includes(`(${files[0]})`)) {
      fail(errors, `${directory}/README.md 未索引当前正文 ${files[0]}`);
    }
  }
}

/** 执行仓库结构、链接、记忆与无 Python 依赖的完整检查。 */
export function validateRepository(errors) {
  let files;
  try {
    files = trackedFiles();
  } catch (error) {
    fail(errors, error.message);
    return;
  }
  validateRequiredFiles(errors);
  validateDailyProjectMemory(errors);
  validateWorkPlanContract(errors);
  validateSkills(errors);
  validateMarkdownLinks(errors, files);
  validateMemoryIndexes(errors);
  for (const error of inspectNoPython(ROOT, { files })) fail(errors, `Harness ${error}`);
}
